API/ev3_sound.c

```diff
@@ -322,7 +322,28 @@ void _playRMDFile(char* pFileName, uint8_t volume, bool loop)
 	}
 }
 
-void _playSoundSamplesFromStream(int fileHandle)
+#define RIFF_HDR_SIZE   44
+#define RIFF_RIFF_SIG   0x52494646
+#define RIFF_WAVE_SIG   0x57415645
+#define RIFF_FMT_SIG    0x666d7420
+#define RIFF_FMT_PCM    0x0100
+#define RIFF_FMT_1CHAN  0x0100
+#define RIFF_FMT_8BITS  0x0800
+#define RIFF_FMT_16BITS 0x1000
+#define RIFF_DATA_SIG   0x64617461
+
+int changeBitDepthFrom16To8 (uint8_t buffer[], int length) {
+	int i, j = 0;
+	for (i = 0; i < length; i += 2) {
+		uint8_t b1 = buffer[i];
+		uint8_t b2 = buffer[i + 1];
+		short tmp = (b1 & 0xff) | ((b2 & 0xff) << 8); // from 2 bytes to short
+		buffer[j++] = ((tmp + 32767) >> 8) & 0xff;

```

Comment:
Signed integer overflow in `tmp` is undefined behavior.

Reply:
If I understood it correctly, the problem appears when if `b1+b2+32767` is greater than `SHORT_MAX`, correct?

If that's correct, would casting `tmp` to an int before doing the sum solve the problem?

```c
short tmp = b1 | (b2 << 8);
buffer[j++] = ((((int)tmp) + 32767) >> 8) & 0xff;
```

Reply:
Never mind, C implicitly converts the `tmp` to an `int` anyway. No overflow here as long as `2*sizeof(short) == sizeof(int)`